src/Text/Pretty/Simple/Internal/OutputPrinter.hs

```diff
@@ -270,141 +327,50 @@ renderOutput (Output _ (OutputCharLit string)) = do
-  (Output nest (OutputOther $ dropWhileEnd isSpace string)):[]
-removeTrailingSpacesInOtherBeforeNewLine (Output nest (OutputOther string):nl@(Output _ OutputNewLine):t) =
-  (Output nest (OutputOther $ dropWhileEnd isSpace string)):nl:removeTrailingSpacesInOtherBeforeNewLine t
-removeTrailingSpacesInOtherBeforeNewLine (h:t) = h : removeTrailingSpacesInOtherBeforeNewLine t
-
--- | If there are two subsequent 'OutputOther' tokens, combine them into just
--- one 'OutputOther'.
---
--- >>> compressOthers [Output 0 (OutputOther "foo"), Output 0 (OutputOther "bar")]
--- [Output {outputNestLevel = NestLevel {unNestLevel = 0}, outputOutputType = OutputOther "foobar"}]
-compressOthers :: [Output] -> [Output]
-compressOthers [] = []
-compressOthers (Output _ (OutputOther string1):(Output nest (OutputOther string2)):t) =
-  compressOthers ((Output nest (OutputOther (string1 `mappend` string2))) : t)
-compressOthers (h:t) = h : compressOthers t
-
--- | In each 'OutputOther' token, compress multiple whitespaces to just one
--- whitespace.
---
--- >>> shrinkWhitespaceInOthers [Output 0 (OutputOther "  hello  ")]
--- [Output {outputNestLevel = NestLevel {unNestLevel = 0}, outputOutputType = OutputOther " hello "}]
-shrinkWhitespaceInOthers :: [Output] -> [Output]
-shrinkWhitespaceInOthers = fmap shrinkWhitespaceInOther
-
-shrinkWhitespaceInOther :: Output -> Output
-shrinkWhitespaceInOther (Output nest (OutputOther string)) =
-  Output nest . OutputOther $ shrinkWhitespace string
-shrinkWhitespaceInOther other = other
-
+-- >>> shrinkWhitespace "  hello    there  "
+-- " hello there "
 shrinkWhitespace :: String -> String
 shrinkWhitespace (' ':' ':t) = shrinkWhitespace (' ':t)
 shrinkWhitespace (h:t) = h : shrinkWhitespace t
 shrinkWhitespace "" = ""
+
+-- | Remove trailing and leading whitespace (see 'Data.Text.strip').
+--
+-- >>> strip "  hello    there  "
+-- "hello    there"
+strip :: String -> String
+strip = dropWhile isSpace . reverse . dropWhile isSpace . reverse
+
+-- | A bidirectional Turing-machine tape:
+-- infinite in both directions, with a head pointing to one element.
+data Tape a = Tape
+  { tapeLeft  :: Stream a -- ^ the side of the 'Tape' left of 'tapeHead'
+  , tapeHead  :: a        -- ^ the focused element
+  , tapeRight :: Stream a -- ^ the side of the 'Tape' right of 'tapeHead'
+  } deriving Show
+-- | Move the head left
+moveL :: Tape a -> Tape a
+moveL (Tape (l :.. ls) c rs) = Tape ls l (c :.. rs)
+-- | Move the head right
+moveR :: Tape a -> Tape a
+moveR (Tape ls c (r :.. rs)) = Tape (c :.. ls) r rs
+
+-- | An infinite list
+data Stream a = a :.. Stream a deriving Show
+-- | Analogous to 'repeat'
+streamRepeat :: t -> Stream t
+streamRepeat x = x :.. streamRepeat x
+-- | Analogous to 'cycle'
+-- While the inferred signature here is more general,
+-- it would diverge on an empty structure
+streamCycle :: NonEmpty a -> Stream a
+streamCycle xs = foldr (:..) (streamCycle xs) xs
```

Comment:
I have no idea how this works. Is it well-tested?

Reply:
It's more or less correct by construction - you end up with e.g.:

```haskell
streamCycle [1,2,3] = 1 :.. ( 2 :.. ( 3 :.. streamCycle [1,2,3] ))
```

Reply:
Thanks @georgefst! That example opened my eyes! :)

Maybe you could include it in the haddocks?!